PushNotifications/PushNotifications/ServerSyncEventHandler.swift

```diff
@@ -0,0 +1,61 @@
+import Foundation
+
+internal class ServerSyncEventHandler {
+    
+    static let serverSyncEventHandlersQueue = DispatchQueue(label: "serverSyncEventHandlersQueue")
+    static var serverSyncEventHandlers = [String: ServerSyncEventHandler]()
+    
+    static func obtain(instanceId: String) -> ServerSyncEventHandler {
+        serverSyncEventHandlersQueue.sync {
+            if let handler = self.serverSyncEventHandlers[instanceId] {
+                return handler
+            } else {
+                let handler = ServerSyncEventHandler()
+                self.serverSyncEventHandlers[instanceId] = handler
+                return handler
+            }
+        }
+    }
+    
+    // used only for testing purposes
+    internal static func destroy(instanceId: String) {
+        serverSyncEventHandlersQueue.sync {
+            self.serverSyncEventHandlers.removeValue(forKey: instanceId)
+        }
+    }
+    
+    internal var userIdCallbacks = Dictionary<String, [(Error?) -> Void]>()
+    internal var stopCallbacks = [() -> Void]()
+    
+    private var interestsChangedDelegates = [() -> InterestsChangedDelegate?]()
+
+    func handleEvent(event: ServerSyncEvent) {
+        DispatchQueue.main.async {
+            switch event {
+            case .InterestsChangedEvent(let interests):
+                self.interestsChangedDelegates.forEach({ delegate in
+                    if let d = delegate() {
+                        d.interestsSetOnDeviceDidChange(interests: interests)
+                    }
+                })
+                
+            case .UserIdSetEvent(let userId, let error):
+                if !(self.userIdCallbacks[userId]?.isEmpty ?? true) {
+                    if let completion = self.userIdCallbacks[userId]?.removeFirst() {
+                        completion(error)
+                    }
+                }
+            case .StopEvent:
+                if !(self.stopCallbacks.isEmpty) {
+                    let completion = self.stopCallbacks.removeFirst()
+                    completion()
+                }
+            }
+        }
+    }
+    
+    func registerInterestsChangedDelegate(_ interestsChangedDelegate: @escaping () -> InterestsChangedDelegate?) {
```

Comment:
This is a delegate for the customer using it, but we need to ensure the delegate calls back the right methods. We don't want to rename things here because it may cause confusion later.